### PR TITLE
Allow repeated measurements in deferred transformer

### PR DESCRIPTION
**Pull request by daxfohl**
Fixes the deferred measurements transformer in the presence of repeated measurements on the same measurement key by maintaining all those measurements as a list and controlling against the correct qubits per the control index.

This change will update diagrams of the output circuits slightly. The ancilla qubits now have indexes to represent the order of the measurement for that key. 

Closes https://github.com/quantumlib/Cirq/issues/5884

### CODE REVIEW

**Comment by tanujkhattar** [comment]
@daxfohl Please resolve the conflicts and I'll be happy to review this PR as well.

**Comment by daxfohl** [comment]
Converting to draft to clean up some merge artifacts

**Comment by daxfohl** [comment]
@tanujkhattar ready

**Comment by daxfohl** [comment]
@tanujkhattar ready

File: cirq-core/cirq/transformers/measurement_transformers_test.py [hunk 1]
```diff
@@ -457,23 +491,23 @@ def test_diagram():
...
+3: ────────────────────┼┼┼@─────────────M────────
+                       ││││
+M('a[0]', q=q(0)): ────X┼┼┼────M('a')────────────
+                        │││    │
+M('a[0]', q=q(2)): ─────X┼┼────M─────────────────
+                         ││
+M('b[0]', q=q(1)): ──────X┼────M('b')────────────
+                          │    │
+M('b[0]', q=q(3)): ───────X────M─────────────────
+                      └────┘
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/measurement_transformers_test.py:510]: Changing diagrams is allowed as per our backwards compatibility guidelines, so this change is fine. But we should update the PR description to explicitly call out that the PR changes diagrams.

**Review by tanujkhattar — APPROVED** [review]

**Pull request merged**